src/rust/engine/process_execution/src/docker.rs

```diff
@@ -362,20 +381,64 @@ impl<'a> super::CommandRunner for CommandRunner<'a> {
           &sandbox_path_in_container
         );
 
+        // Obtain ID of the base container in which to run the execution for this process.
+        let container_id = {
+          let image = match &req.execution_strategy {
+            ProcessExecutionStrategy::Docker(image) => Ok(image),
+            pes => Err(format!(
+              "The Docker execution strategy was: {pes:?}, but the Docker CommandRunner was used."
+            )),
+          }?;
+
+          self
+            .container_cache
+            .container_id_for_image(image, &req.platform, &context.build_id)
+            .await?
+        };
+
         // Prepare the workdir.
-        // DOCKER-NOTE: The input root will be bind mounted into the container.
-        let exclusive_spawn = prepare_workdir(
-          workdir.path().to_owned(),
-          &req,
-          req.input_digests.input_files.clone(),
-          self.store.clone(),
-          self.executor.clone(),
-          &self.named_caches,
-          &self.immutable_inputs,
-          Some(Path::new(NAMED_CACHES_BASE_PATH_IN_CONTAINER)),
-          Some(Path::new(IMMUTABLE_INPUTS_BASE_PATH_IN_CONTAINER)),
-        )
-        .await?;
+        match self.strategy {
+          DockerStrategy::Mount => {
+            prepare_workdir(
+              workdir.path().to_owned(),
+              &req,
+              req.input_digests.input_files.clone(),
+              &self.store,
+              &self.named_caches,
+              &self.immutable_inputs,
+              Some(Path::new(NAMED_CACHES_BASE_PATH_IN_CONTAINER)),
+              Some(Path::new(IMMUTABLE_INPUTS_BASE_PATH_IN_CONTAINER)),
+            )
+            .await?;
+          }
+          DockerStrategy::Pipe => {
+            // NB: For now, we continue to materialize files into a bind mount, but we do so from
+            // within the container. This allows us to capture outputs without an additional
```

Comment:
This seems logically very likely. In my reading this is a known bug we're just waiting for to happen without calling the shot in an issue, maybe do that and we get to work on it? This still feels like shipping known (slightly less) broken software.

Reply:
It seems possible, yes. But since we haven't observed it yet, I'm gambling... in the interest of saving time and complexity (I hope).

Reply:
With my user hat on I definitely do not appreciate you gambling and not letting me know. I hope this will make a release announcement / acknowledgement - basically macOS Docker integration is known likely broken and you may see X.

Reply:
If you want a semi-reliable way manual way to expose the bug, clone https://github.com/tdyas/pants-go-testing and then run `./pants package race:racy_docker` repeatedly.

Reply:
(although "race" there refers to the Go data race detector, not this bug)

Reply:
Mm, missed your comment before merging @tdyas : but thanks, will try it.